core/src/main/java/com/linecorp/armeria/server/VirtualHostBuilder.java

```diff
@@ -960,29 +1015,123 @@ VirtualHost build(VirtualHostBuilder template) {
                                requestContentPreviewerFactory, responseContentPreviewerFactory,
                                accessLogWriter, shutdownAccessLogWriterOnStop);
 
-        SslContext sslContext = this.sslContext != null ? this.sslContext : template.sslContext;
-        final boolean tlsSelfSigned = this.tlsSelfSigned != null ? this.tlsSelfSigned : template.tlsSelfSigned;
-        if (sslContext == null && tlsSelfSigned) {
-            try {
-                final SelfSignedCertificate ssc = new SelfSignedCertificate(defaultHostname);
-                tls(ssc.certificate(), ssc.privateKey());
+        SslContext sslContext = null;
+        boolean releaseSslContextOnFailure = false;
+        try {
+            // Build a new SslContext or use a user-specified one for backward compatibility.
+            if (this.sslContext != null) {
                 sslContext = this.sslContext;
-            } catch (Exception e) {
-                throw new RuntimeException("failed to create a self signed certificate", e);
+            } else if (sslContextBuilder != null) {
```

Comment:
What happens only if `tlsCustomizer(Consumer<? super SslContextBuilder> tlsCustomizer)` is called?
Should we raise an exception if `tlsCustomizer()` is called before `sslContextBuilder` or `tlsSelfSigned` is set ?

Reply:
Because the following should work:

```java
sb.tlsCustomizer(...)
  .tlsSelfSigned()
  ...
```

we will have to do something at `build()`. Let me raise an exception if customizers were specified without `tls()` or `tlsSelfSigned()`.

Reply:
Fixed